src/uct/ib/mlx5/ib_mlx5.inl

```diff
@@ -72,28 +84,155 @@ uct_ib_mlx5_update_db_cq_ci(uct_ib_mlx5_cq_t *cq)
-uct_ib_mlx5_poll_cq(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq)
+uct_ib_mlx5_iface_cqe_decompress_continue(uct_ib_mlx5_cq_t *cq)
 {
+    uct_ib_mlx5_cq_decompress_t *cq_decompress
+                                      = &cq->cq_decompress;
+    uint8_t id                        = cq_decompress->mini_arr_idx;
+    struct mlx5_cqe64 *title          = &cq_decompress->title;
+    uct_ib_mlx5_mini_cqe8_t *mini_cqe = &cq_decompress->mini_arr[id % 8];
     struct mlx5_cqe64 *cqe;
-    unsigned cqe_index;
-    uint8_t op_own;
 
-    cqe_index = cq->cq_ci;
-    cqe       = uct_ib_mlx5_get_cqe(cq, cqe_index);
-    op_own    = cqe->op_own;
+    uct_ib_mlx5_iface_cqe_decompress_fill_unique_fields(title, mini_cqe,
+                                                        cq_decompress);
 
-    if (ucs_unlikely(uct_ib_mlx5_cqe_is_hw_owned(op_own, cqe_index, cq->cq_length))) {
-        return NULL;
-    } else if (ucs_unlikely(op_own & UCT_IB_MLX5_CQE_OP_OWN_ERR_MASK)) {
+    cq_decompress->left--;
+    cq_decompress->mini_arr_idx++;
+    cq->cq_ci++;
+
+    if (cq_decompress->left != 0) {
+        if (cq_decompress->mini_arr_idx % 8 == 0) {
+            /* Get the next mini_cqe array */
+            cqe = uct_ib_mlx5_get_cqe(cq, cq->cq_ci);
+            memcpy(&cq_decompress->mini_arr, cqe, sizeof(*cqe));
+        }
+        /* Update opcode and CQE format in the next CQE buffer
+         * (title shouldn't be updated).
+         * CQE format is used for simplifying the compressed CQE detection
+         * during the poll
+         */
+        cqe         = uct_ib_mlx5_get_cqe(cq, cq->cq_ci);
+        cqe->op_own = cq_decompress->title_op_own;
+    }
+
+    return title;
+}
+
+
+/**
+ * Initialize the CQE decompression related entities and decompress the first CQE.
+ *
+ * @param title The CQE that contains the title of the compression block.
+ * @param cq    CQ that contains the title.
+ * 
+ * @return First decompressed cqe.
+ */
+static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
+uct_ib_mlx5_iface_cqe_decompress_start(struct mlx5_cqe64 *title,
+                                       uct_ib_mlx5_cq_t *cq)
+{
+    uct_ib_mlx5_cq_decompress_t *cq_decompress
+                                = &cq->cq_decompress;
+    struct mlx5_cqe64 *mini_cqe = uct_ib_mlx5_get_cqe(cq, cq->cq_ci + 1);
+
+    cq_decompress->block_size   = ntohl(title->byte_cnt);
+    cq_decompress->left         = cq_decompress->block_size;
+    cq_decompress->wqe_counter  = ntohs(title->wqe_counter);
+    cq_decompress->mini_arr_idx = 0;
+    cq_decompress->title_op_own = title->op_own;
+    memcpy(&cq_decompress->title,    title,     sizeof(struct mlx5_cqe64));
+    memcpy(&cq_decompress->mini_arr, mini_cqe,  sizeof(struct mlx5_cqe64));
+
+    UCT_IB_MLX5_CQE_FORMAT_CLEAR(cq_decompress->title.op_own);
+    return uct_ib_mlx5_iface_cqe_decompress_continue(cq);
+}
+
+
+static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
+uct_ib_mlx5_iface_cqe_decompress(struct mlx5_cqe64 *cqe,
```

Comment:
IMO better extend uct_ib_mlx5_check_completion() to also handle compresses CQEs (and move some asserts accordingly)